dask_glm/algorithms.py

```diff
@@ -138,6 +140,83 @@ def gradient_descent(X, y, max_iter=100, tol=1e-14, family=Logistic, **kwargs):
     return beta
 
 
+@normalize
+def sgd(X, y, epochs=100, tol=1e-3, family=Logistic, batch_size=64,
+        initial_step=1e-4, callback=None, average=True):
+    """Stochastic Gradient Descent.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+    y : array-like, shape (n_samples,)
+    epochs : int, float
+        maximum number of passes through the dataset
+    tol : float
+        Maximum allowed change from prior iteration required to
+        declare convergence
+    batch_size : int
+        The batch size used to approximate the gradient. Larger batch sizes
+        will approximate the gradient better.
+    initial_step : float
+        The initial step size. The step size is decays like 1/k.
+    callback : callable
+        A callback to call every iteration that accepts keyword arguments
+        `X`, `y`, `beta`, `grad`, `nit` (number of iterations) and `family`
+    average : bool
+        To average the parameters found or not. See [1]_.
+    family : Family
+
+    Returns
+    -------
+    beta : array-like, shape (n_features,)
+
+    .. _1: https://en.wikipedia.org/wiki/Stochastic_gradient_descent#Averaging
+    """
+    gradient = family.gradient
+    n, p = X.shape
+    if np.isnan(n):
+        raise ValueError('SGD needs shape information to allow indexing. '
+                         'Possible by passing a computed array in (`X.compute()` '
+                         'or `X.values.compute()`), then doing using '
+                         '`dask.array.from_array ')
```

Comment:
I've allowed passing a keyword arg in to get the number of examples. I print an error message if it's not present, and provide helpful examples with a distributed dataframe (implementation is at [algorithms.py#_get_n][1]).

Is this more of what you'd like to see? I debated going to a dataframe and computing the length from that, but this option was cleanest and allowed the user to use outside information.

[1]:https://github.com/stsievert/dask-glm/blob/b8757350a20eef48d0333f044e97a3bc8fb46e8f/dask_glm/algorithms.py#L145-L153

Reply:
In general this is something that we should probably push upstream into dask.array proper.  It looks like there is an upstream issue here: https://github.com/dask/dask/issues/3293

Short term if you're experimenting then I say that we just go with this and move on.  Long term I don't think that it would be reasonable to ask users to provide this information, but if we're ok not merging this implementation (which given performance issues seems wise) then it might be best to just blow past it for now.

Reply:
Leaving to the upstream issue sounds best to me.